doc/programming/guidelines.rst

```diff
@@ -0,0 +1,291 @@
+see if you need to improve performance, and finally find the primary
+bottlenecks of needed.
+
+.. _performance_runtime:
+
+Runtime Performance
+-------------------
+
+Among other things, runtime performance is affected by:
+    - additional C++ code that is generated by the Spicy compiler
+    - use of expensive Spicy constructs
+    - poorly managed objects hogging memory (both on stack and heap)
+
+Below are some strategies to make the runtime performance as efficient as possible.
+
+.. rubric:: Avoid declaring public units
+
+The Spicy compiler generates additional code for ``public`` units,
+thereby increasing both compile and execution times. Defining units as
+``public`` is only required for top-level units where parsing starts
+(i.e., with Zeek: the units that you enter into your EVT file). The
+attribute can be omitted otherwise.
+
+.. rubric:: Turn unreferenced vectors into anonymous fields
+
+Implement vectors that don't need to be referenced from your code as
+:ref:`anonymous fields <anonymous_fields>` (i.e., do not provide a
+name for the field). Normal, named fields store the whole vector
+accumulating all parsed elements over the lifetime of the defining
+unit; whereas anonymous vector field forgo storage, because nobody
+could access the fields anyways.
+
+The most common pitfall here is top-level units that parse a sequence
+of PDUs:
+
+.. code::
+
+    public type MyPDUs = unit {
+        msgs: Message[]; # accumulates all messages until end of session
+        : Message[];     # anonymous field, no accumulation
+    };
+
+.. rubric::  Skip unused fields
+
+Don't parse data that is not needed. Parsing data into a field will
+cause a copy of that data into a dedicated memory location. Instead,
+use the :ref:`skip keyword <skip>` to discard data that isn't of
+interest, which leads the compiler to generate optimized code for many
+field types, including in particular ``bytes``, literals, and
+generally fields of fixed size. Examples:
+
+.. code::
+
+    public type Message = unit {
+        unused1: uint32;            # parse and store
+        unused2: skip uint32;       # skip over 4 bytes
+
+        unused3: bytes &eod;        # extract and store remaining data
+        unused4: skip bytes &eod;   # skip over remaining data
+    };
+
+.. rubric::  Avoid Spicy strings and string manipulation
+
+Use Spicy :ref:`strings <type_string>` sparsely in your analyzer, and
+stick to :ref:`bytes <type_bytes>` instead where you can. Typically,
+you would convert ``bytes`` to strings as late as possible just before
+presentation to the user (including passing to Zeek events). Strings
+are never needed for the parsing process itself.
+
+As a corollary, avoid Spicy string manipulation. Always
+manipulate/concatenate ``bytes`` and convert only the final result to
+Spicy strings, probably in a ``%done`` hook or an EVT event. In
+particular, format strings come with a significant overhead. In other
```

Comment:
Nit: personally triggered by "overhead", maybe this?

```suggestion
particular, format strings are not cheap to compute. In other
```

Reply:
Merging the two to stay with active voice. :-)